src/WinUsb/WinUsb.Helpers.cs

```diff
@@ -0,0 +1,137 @@
+        /// </returns>
+        public static unsafe ValueTask<int> WinUsb_ReadPipeAsync(SafeUsbHandle interfaceHandle, byte pipeID, Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            var overlapped = new WinUsbOverlapped(interfaceHandle, pipeID, buffer);
+            var nativeOverlapped = overlapped.Pack();
+
+            cancellationToken.Register(overlapped.Cancel);
+
+            using (var memoryHandle = buffer.Pin())
+            {
+                if (WinUsb_ReadPipe(
+                    interfaceHandle,
+                    pipeID,
+                    (byte*)memoryHandle.Pointer,
+                    buffer.Length,
+                    out int lengthTransferred,
+                    nativeOverlapped))
+                {
+                    overlapped.Unpack();
+                    return new ValueTask<int>(lengthTransferred);
+                }
+                else
+                {
+                    var error = (Win32ErrorCode)Marshal.GetLastWin32Error();
+
+                    if (error == Win32ErrorCode.ERROR_IO_PENDING)
+                    {
+                        return new ValueTask<int>(overlapped.Completion);
+                    }
+                    else
+                    {
+                        overlapped.Unpack();
+
+#if NET45
+                        return new ValueTask<int>(Task.Run(new Func<int>(() => throw new Win32Exception(error))));
+#else
+                        return new ValueTask<int>(Task.FromException<int>(new PInvoke.Win32Exception(error)));
+#endif
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously writes data to a pipe.
+        /// </summary>
+        /// <param name="interfaceHandle">
+        /// An opaque handle to the interface that contains the endpoint with which the pipe is associated.
+        /// </param>
+        /// <param name="pipeID">
+        /// PipeID corresponds to the bEndpointAddress field in the endpoint descriptor.
+        /// </param>
+        /// <param name="buffer">
+        /// A caller-allocated buffer that contains the data to write.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// A <see cref="CancellationToken"/> which can be used to cancel the asynchronous operation.
+        /// </param>
+        /// <returns>
+        /// A <see cref="ValueTask"/> which represents the asynchronous operation, and returns the number of bytes transferred.
+        /// </returns>
+        public static unsafe ValueTask<int> WinUsb_WritePipeAsync(SafeUsbHandle interfaceHandle, byte pipeID, Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            var overlapped = new WinUsbOverlapped(interfaceHandle, pipeID, buffer);
+            var nativeOverlapped = overlapped.Pack();
+
+            cancellationToken.Register(overlapped.Cancel);
+
+            using (var memoryHandle = buffer.Pin())
```

Comment:
This `memoryHandle` will be disposed (and `buffer` unpinned) as soon as this block exits, which will happen *before* the async operation has completed. That would likely cause data corruption.

Instead, I think you have to pass this handle into your `overlapped` object so that it will dispose it as it sets `Completion` to its completed state. Just be careful to consider all exception paths so that it is disposed of in all cases.